ui/app/hooks/useTokenTracker.js

```diff
@@ -0,0 +1,95 @@
+import { useState, useEffect, useRef, useCallback } from 'react'
+import TokenTracker from '@metamask/eth-token-tracker'
+import { isEqual } from 'lodash'
+import { useSelector } from 'react-redux'
+import { getCurrentNetwork, getSelectedAddress } from '../selectors'
+import { getTokens } from '../ducks/metamask/metamask'
+
+
+export function useTokenTracker () {
+  const network = useSelector(getCurrentNetwork)
+  const userAddress = useSelector(getSelectedAddress)
+  // use `isEqual` comparison function because the token array is serialized
+  // from the background so it has a new reference with each background update,
+  // even if the tokens haven't changed
+  const tokens = useSelector(getTokens, isEqual)
+
+  const [loading, setLoading] = useState(() => tokens?.length >= 0)
+  const [tokensWithBalances, setTokensWithBalances] = useState([])
+  const [error, setError] = useState(null)
+  const tokenTracker = useRef(null)
+
+  const updateBalances = useCallback((tokensWithBalances) => {
+    setLoading(false)
+    setTokensWithBalances(tokensWithBalances)
+    setError(null)
+  }, [])
+
+  const showError = useCallback((error) => {
+    setError(error)
+    setLoading(false)
+  }, [])
+
+  const teardownTracker = useCallback(() => {
+    if (tokenTracker.current) {
+      tokenTracker.current.stop()
+      tokenTracker.current.removeAllListeners('update')
+      tokenTracker.current.removeAllListeners('error')
+      tokenTracker.current = null
+    }
+  }, [])
+
+  const buildTracker = useCallback((address, tokenList) => {
+    // clear out previous tracker, if it exists.
+    teardownTracker()
+    if (!tokenTracker.current) {
+      tokenTracker.current = new TokenTracker({
+        userAddress: address,
+        provider: global.ethereumProvider,
+        tokens: tokenList,
+        pollingInterval: 8000,
+      })
+
+      tokenTracker.current.on('update', updateBalances)
+      tokenTracker.current.on('error', showError)
+      tokenTracker.current.updateBalances()
+    }
```

Comment:
Nit: This condition should always be `true`, since it immediately follows a teardown.

```suggestion
    tokenTracker.current = new TokenTracker({
      userAddress: address,
      provider: global.ethereumProvider,
      tokens: tokenList,
      pollingInterval: 8000,
    })

    tokenTracker.current.on('update', updateBalances)
    tokenTracker.current.on('error', showError)
    tokenTracker.current.updateBalances()
```